File: tauri/src-tauri/src/llm/openrouter.rs
use crate::llm::{LlmProvider, Message, LlmEvent, Usage};
use async_trait::async_trait;
use futures_util::{Stream, StreamExt};
use reqwest::Client;
use serde_json::Value;
use std::pin::Pin;

pub struct OpenRouter {
    api_key: String,
    base_url: String,
    client: Client,
}

impl OpenRouter {
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            base_url: "https://openrouter.ai/api/v1".to_string(),
            client: Client::new(),
        }
    }
}

#[async_trait]
impl LlmProvider for OpenRouter {
    async fn stream_chat(
        &self,
        model: &str,
        messages: Vec<Message>,
    ) -> Result<Pin<Box<dyn Stream<Item = LlmEvent> + Send>>, String> {
        if self.api_key.is_empty() {
            return Err("OpenRouter API Key not found".to_string());
        }

        let response = self.client
            .post(format!("{}/chat/completions", self.base_url))
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .header("HTTP-Referer", "https://github.com/methil-mods/mosaic")
            .header("X-Title", "Mosaic")
            .json(&serde_json::json!({
                "model": model,
                "messages": messages,
                "stream": true,
                "stream_options": { "include_usage": true }
            }))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let stream = response.bytes_stream().map(|result| {
            match result {
                Ok(bytes) => {
                    let chunk = String::from_utf8_lossy(&bytes);
                    let mut events = Vec::new();
                    
                    for line in chunk.lines() {
                        let trimmed = line.trim();
                        if trimmed.is_empty() || !trimmed.starts_with("data: ") {
                            continue;
                        }
                        
                        let data = &trimmed[6..];
                        if data == "[DONE]" {
                            // We handle complete via accumulation or just wait for finish_reason
                            continue;
                        }
                        
                        if let Ok(json) = serde_json::from_str::<Value>(data) {
                            if let Some(usage) = json.get("usage") {
                                if let Ok(u) = serde_json::from_value::<Usage>(usage.clone()) {
                                    events.push(LlmEvent::Usage { usage: u });
                                }
                            }
                            
                            if let Some(choices) = json.get("choices").and_then(|c| c.as_array()) {
                                if let Some(choice) = choices.get(0) {
                                    if let Some(delta) = choice.get("delta") {
                                        if let Some(content) = delta.get("content").and_then(|c| c.as_str()) {
                                            events.push(LlmEvent::Token { token: content.to_string() });
                                        }
                                    }
                                }
                            }
                        }
                    }
                    events
                }
                Err(e) => vec![LlmEvent::Error { message: e.to_string() }],
            }
        })
        .flat_map(futures_util::stream::iter);

        Ok(Box::pin(stream))
    }

    async fn fetch_models(&self) -> Result<Vec<String>, String> {
        // Not strictly needed for the port right now, but good to have
        Ok(vec![])
    }
}

File: cli/mosaic_cli/framework/llm/openrouter.py
import httpx
import json
from typing import AsyncIterable, List, Dict, Any
from .base import LlmProvider

class OpenRouter(LlmProvider):
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://openrouter.ai/api/v1"

    async def stream_chat(self, model: str, messages: List[Dict[str, str]]) -> AsyncIterable[Dict[str, Any]]:
        if not self.api_key:
            yield {"type": "error", "message": "OpenRouter API Key not found"}
            return

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://github.com/methil-mods/mosaic",
            "X-Title": "Mosaic",
        }
        
        payload = {
            "model": model,
            "messages": messages,
            "stream": True,
        }

        async with httpx.AsyncClient() as client:
            try:
                async with client.stream("POST", f"{self.base_url}/chat/completions", headers=headers, json=payload, timeout=60.0) as response:
                    if response.status_code != 200:
                        yield {"type": "error", "message": f"OpenRouter HTTP {response.status_code}: {await response.aread()}"}
                        return

                    async for line in response.aiter_lines():
                        line = line.strip()
                        if not line or not line.startswith("data: "):
                            continue

                        data = line[6:]
                        if data == "[DONE]":
                            continue

                        try:
                            chunk = json.loads(data)
                            if "usage" in chunk:
                                yield {"type": "usage", "data": chunk["usage"]}
                            
                            delta = chunk.get("choices", [{}])[0].get("delta", {})
                            if "content" in delta:
                                yield {"type": "token", "data": delta["content"]}
                        except json.JSONDecodeError:
                            continue
            except Exception as e:
                yield {"type": "error", "message": str(e)}

    async def fetch_models(self) -> List[str]:
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(f"{self.base_url}/models", headers=headers)
                if response.status_code == 200:
                    data = response.json()
                    return [m["id"] for m in data.get("data", [])]
            except:
                pass
        return []
